src/NuGet.Core/NuGet.Common/ActivityCorrelationContext.cs

```diff
@@ -0,0 +1,84 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+
+#if DNXCORE50
+using System.Threading;
+#else
+using System.Runtime.Remoting.Messaging;
+#endif
+
+namespace NuGet.Common
+{
+    /// <summary>
+    /// Ambient context contains properties pertaining to a current activity.
+    /// Single activity engages multiple method calls at different layers. 
+    /// Sometimes it's necessary to identify separate calls belonging to the same activity if shared state is needed.
+    /// </summary>
+    public class ActivityCorrelationContext : IDisposable
+    {
+        private static readonly string DataSlotId = "NuGet.ActivityCorrelationContext";
+
+#if DNXCORE50
+        private static readonly AsyncLocal<ActivityCorrelationContext> _instance = new AsyncLocal<ActivityCorrelationContext>();
+#endif
+
+        private static readonly ActivityCorrelationContext _globalInstance = new ActivityCorrelationContext
+        {
+            CorrelationId = default(Guid).ToString()
+        };
+
+        /// <summary>
+        /// Represents activity unique ID.
+        /// </summary>
+        public string CorrelationId { get; private set; }
+
+        /// <summary>
+        /// Returns ambient context value or global instance if not set earlier.
+        /// </summary>
+        public static ActivityCorrelationContext Current
+        {
+            get
+            {
+#if DNXCORE50
+                return _instance.Value ?? _globalInstance;
+#else
+                var context = CallContext.LogicalGetData(DataSlotId);
```

Comment:
`AsyncLocal<T>` is a new class available since .NET 4.6. As the fix initially targeted 3.4.x release I had to provide alternative solution using good old `CallContext`. 

I'll go ahead and re-write it to use `AsyncLocal` everywhere for dev.